File: apps/render-ios-playground/render-ios-playground/Application/Services/ScenarioService.swift
import UIKit

/// Application service that orchestrates schema operations
class ScenarioService {
    private let fetchScenarioUseCase: FetchScenarioUseCase
    
    init(fetchScenarioUseCase: FetchScenarioUseCase) {
        self.fetchScenarioUseCase = fetchScenarioUseCase
    }
    
    func fetchScenario(from url: URL) async throws -> Scenario {
        guard let scenario = try? await fetchScenarioUseCase.execute() else {
            throw ApplicationError.scenarioFetchFailed("No scenario recieved")
        }
        
        return scenario
    }
}

File: apps/render-android-playground/app/src/main/java/com/max_mrtnv/render_android_playground/application/services/ScenarioService.kt
package com.max_mrtnv.render_android_playground.application.services

import androidx.compose.runtime.Composable
import com.max_mrtnv.render_android_playground.application.errors.ApplicationError
import com.max_mrtnv.render_android_playground.application.usecases.FetchScenarioUseCase
import com.max_mrtnv.render_android_playground.application.usecases.RenderComponentUseCase
import com.max_mrtnv.render_android_playground.domain.entities.Scenario

/**
 * Application service that orchestrates scenario operations
 */
class ScenarioService(
    private val fetchScenarioUseCase: FetchScenarioUseCase,
    private val renderComponentUseCase: RenderComponentUseCase
) {
    suspend fun fetchScenario(): Scenario {
        return fetchScenarioUseCase.execute() 
            ?: throw ApplicationError.SchemaFetchFailed("No scenario received")
    }
    
    suspend fun fetchScenario(url: String): Scenario {
        return fetchScenarioUseCase.execute(url)
            ?: throw ApplicationError.SchemaFetchFailed("No scenario received")
    }
    
    @Composable
    fun renderScenario(scenario: Scenario): Boolean {
        return renderComponentUseCase.execute(scenario)
    }
}
